File: cpp/observer/Subject.cpp
#include "Subject.hpp"

Subject::Subject() {
  this->state = 0;
}

Subject::~Subject() {

}

void Subject::attach(View *a) {
  this->observers.push_back(a);
}

void Subject::detach(View *a) {
  /* TODO: Implement this method */
}

void Subject::notify() {
  auto it = this->observers.begin();
  auto end = this->observers.end();

  for (View *v : this->observers)
    v->update(this->state);
}

void Subject::setState(int s) {
  this->state = s;
  this->notify();
}

File: js/node_chat/backend/Subject.js
const Logger = require('./Logger.js');


class Subject {

  constructor() {
    this.logger = new Logger('Subject.js');
    this.observers = [];
    this._subject_state = null;
  }

  attach(observer) {
    this.observers.push(observer);
    this.logger.info(`Adding new observer: ${observer.name}`);
  }

  detach(observer) {
    this.observer.pop(observer);
  }

  notify() {
    this.observers.forEach((obj, id) => {
      obj.update(this._subject_state);
      this.logger.info(`Notifying obj: ${obj.name}. State: ${this._subject_state}`);
    });
  }

  setSubjectState(state) {
    this._subject_state = state;
    this.notify();
  }
}

module.exports = Subject;